test/livebook_web/live/hub/new/enterprise_component_test.exs

```diff
@@ -22,9 +24,13 @@ defmodule LivebookWeb.Hub.New.EnterpriseComponentTest do
         }
       })
 
-      assert view
-             |> element("#connect")
-             |> render_click() =~ "Add Hub"
+      view
+      |> element("#connect")
+      |> render_click()
+
+      Process.sleep(50)
```

Comment:
I'm not happy with that, if I don't add this sleep time, the assert below fails because the view didn't update yet.

I've tried the Chris McCord suggestion from [here](https://elixirforum.com/t/testing-liveview-that-uses-send-update/30286/2).

Reply:
We no longer make the update on `handle_event("connect", ...)`, but only later when we get the enterprise reply. In some cases we can make the test wait for the relevant event (e.g. by subscribing to a pubsub). Here the websocket message is sent directly to the LV process and we don't persist anything yet, so I'm not sure if there's any event we can await on :<

Reply:
I am trying to think what is the best design here. Ideally, I would try to make this particular connection synchronous:

1. Do a WebSocket upgrade to the desired endpoint
2. See if the status code is 200 and if so, it is connected

Of course, if we are using Mint, we don't want to run this inside the LiveView process, so we can do it inside a task, something like this:

    case Task.async(fn -> ... end) |> Task.yield(5000) |> Task.shutdown(:kill) do
      nil -> "timed out"
      {:exit, _} -> "error while connecting"
      {:ok, %{... 200 ...}} -> "valid"
      {:error, %{... _ ...}} -> "invalid"
    end

The issue is that, once the Enterprise Hub is saved, we will most likely want to spawn a WebSocket connection to the Hub whenever Livebook starts. This particular connection, should stay always running, and automatically connect on it. If it temporarily drops, we want to automatically reconnect, and so on.

In other  words, I think we will need both sync and async mode. The good news is that we will be able to share most of the logic I think. My suggestion is to add a sync mode, keeping the GenServer as is, and then refactor the GenServer to be fully async (i.e. the GenServer should receive all connection parameters on init and attempt to establish the connection asynchronously). Perhaps, for the GenServer, we should consider using https://github.com/elixir-ecto/connection/?